File: 210-course-schedule-ii/solution.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        self.graphs = self.build_graph(numCourses, prerequisites)
        
        # define variables
        self.visited = [False] * numCourses
        self.onpath = [False] * numCourses
        self.cycle = False

        self.poster_order = []

        for node in range(numCourses):
            self.dfs(node)
        
        if self.cycle:
            return []
        
        return reversed(self.poster_order)

    def build_graph(self, numCourses, prerequisites):
        graphs = [list() for _ in range(numCourses)]
        for edge in prerequisites:
            _from, _to = edge[1], edge[0]
            graphs[_from].append(_to)
        return graphs

    def dfs(self, src):
        if self.onpath[src]:
            self.cycle = True
        
        if self.visited[src] or self.cycle:
            return
        
        self.visited[src] = True
        self.onpath[src] = True
        
        for n in self.graphs[src]:
            self.dfs(n)
        
        self.poster_order.append(src)
        self.onpath[src] = False

File: 210-course-schedule-ii/solution.cpp
class Solution {
public:
    vector<vector<int>> graphs;
    vector<bool> visited;
    vector<bool> onpath;
    vector<int> poster_order;
    bool cycle;

    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        graphs = build_graph(numCourses, prerequisites);

        visited = vector<bool>(numCourses, false);
        onpath = vector<bool>(numCourses, false);
        cycle = false;

        for (int i = 0; i < numCourses; i++) {
            dfs(i);
        }

        if (cycle) {
            return vector<int>();
        }

        reverse(poster_order.begin(), poster_order.end());
        return poster_order;
    }

    vector<vector<int>> build_graph(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graphs(numCourses);
        for (const auto& edge : prerequisites) {
            int from = edge[1], to = edge[0];
            graphs[from].push_back(to);
        }
        return graphs;
    }

    void dfs(int src) {
        if (onpath[src]) {
            cycle = true;
        }

        if (visited[src] || cycle) {
            return;
        }

        visited[src] = true;
        onpath[src] = true;

        for (const auto& n : graphs[src]) {
            dfs(n);
        }

        poster_order.push_back(src);
        onpath[src] = false;
    }
};
